### PR TITLE
Add nanoseconds to timestamp to make it RFC3339Nano

### LINKED ISSUE #471: Support RFC3339Nano
We support RFC3339 but not the nanoseconds in the logs.

### PR DESCRIPTION
**Pull request by mrunalp**
Fixes #471 
Signed-off-by: Mrunal Patel <mpatel@redhat.com>

### CODE REVIEW

**Comment by jawnsy** [comment]
LGTM. Dunno why we need nanoseconds though :laughing:

**Comment by mrunalp** [comment]
@jawnsy Gotta satisfy the CRI :)

**Comment by mrunalp** [comment]
@cyphar @sameo @runcom PTAL

**Comment by runcom** [comment]
LGTM

File: conmon/conmon.c [hunk 1]
```diff
@@ -93,28 +93,32 @@ static GOptionEntry entries[] =
...
 	struct tm *tm;
+	struct timespec ts;
 	char off_sign = '+';
 	int off, len, err = -1;
 
 	if ((tm = localtime(&now)) == NULL)
 		return err;
 
+	if (clock_gettime(CLOCK_REALTIME, &ts) < 0)
+		return err;
```
Comment (mikebrow) [inline conmon/conmon.c]: if you can't get the realtime clock i'd just continue with ts.tv_nsec = 0...

Comment (mrunalp) [inline conmon/conmon.c]: Updated.

File: conmon/conmon.c [hunk 2]
```diff
@@ -93,28 +93,39 @@ static GOptionEntry entries[] =
...
 	char off_sign = '+';
 	int off, len, err = -1;
+	int rc;
 
 	if ((tm = localtime(&now)) == NULL)
 		return err;
 
+	if ((rc = clock_gettime(CLOCK_REALTIME, &ts)) < 0) {
+		/* If CLOCK_REALTIME is not supported, we set nano seconds to 0 */
+		if (rc == EINVAL) {
```
Comment (mikebrow) [inline conmon/conmon.c]: should be errno == EINVAL I think...

Comment (cyphar) [inline conmon/conmon.c]: Should be `errno`.

File: conmon/conmon.c [hunk 3]
```diff
@@ -93,28 +93,39 @@ static GOptionEntry entries[] =
...
 	struct tm *tm;
+	struct timespec ts;
 	char off_sign = '+';
 	int off, len, err = -1;
+	int rc;
 
 	if ((tm = localtime(&now)) == NULL)
 		return err;
 
+	if ((rc = clock_gettime(CLOCK_REALTIME, &ts)) < 0) {
```
Comment (cyphar) [inline conmon/conmon.c]: We don't need `rc`. Just do `clock_gettime(...) < 0`.

**Comment by mrunalp** [comment]
Updated.

**Comment by mrunalp** [comment]
One more change on the way. Reducing to just one call to get time.

File: conmon/conmon.c (hunk 3 above)
Comment (mikebrow) [inline conmon/conmon.c]: if(clock_gettime(...)) :-) but who's counting chars the optimizer will figure it out.

**Comment by mikebrow** [comment]
fancy

**Comment by cyphar** [comment]
LGTM, so fancy. :wink:

**Comment by mrunalp** [comment]
Alright pushed that as well.

**Comment by sameo** [comment]
LGTM

File: conmon/conmon.c [hunk 4]
```diff
@@ -93,28 +93,38 @@ static GOptionEntry entries[] =
...
+			ts.tv_nsec = 0;
+		} else {
+			return err;
+		}
+	}
+
+	if ((tm = localtime(&ts.tv_sec)) == NULL)
 		return err;
 
+
```
Comment (mikebrow) [inline conmon/conmon.c]: blank line.. nit

Haven't run it but presuming the process / local time zone is running in gmt  ..

Comment (cyphar) [inline conmon/conmon.c]: `tm->tm_gmtoff` presumably takes care of this, though this is actually missing from my man page...

Comment (mikebrow) [inline conmon/conmon.c]: yeah not sure what the "format" is supposed to be exactly.. for example some of the docker stuff uses:
RFC3339Nano = "2006-01-02T15:04:05.999999999Z07:00"

Comment (mrunalp) [inline conmon/conmon.c]: I tested this with k8s. This is what it expects ;)

Comment (mikebrow) [inline conmon/conmon.c]: docker events
2017-04-25T18:51:57.337681796-05:00 container create 13ac2a172e18b25e89a9b7924f71c54b6f1d5efe6e0e985e11a29dc42fe4a926 (image=hello-world, name=nervous_franklin)

Comment (cyphar) [inline conmon/conmon.c]: @mikebrow Yeah, that format is what's being output by this code. `gmtoff` generates the final component (the UTC offset).

File: conmon/conmon.c [hunk 5]
```diff
@@ -93,28 +93,38 @@ static GOptionEntry entries[] =
...
+	if (clock_gettime(CLOCK_REALTIME, &ts) < 0) {
+		/* If CLOCK_REALTIME is not supported, we set nano seconds to 0 */
+		if (errno == EINVAL) {
+			ts.tv_nsec = 0;
+		} else {
+			return err;
+		}
+	}
+
+	if ((tm = localtime(&ts.tv_sec)) == NULL)
```
Comment (TomSweeneyRedHat) [inline conmon/conmon.c]: Nit.  FWIW, I don't think this err will ever be returned as is.  I think worse case is tm would be set to the epoch time (1969-12-31T19:00:01 when run on the east coast) from localtime.  I don't know if you want to test for that or not and return -1 if that is the value for tm.

Comment (mrunalp) [inline conmon/conmon.c]: @TomSweeneyRedHat Not sure what you mean. We are returning an error from the function if localtime fails.

**Pull request merged**